lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py

```diff
@@ -1,31 +1,226 @@
 import pytest
 
 
-def test_pbench_list_tools_help():
-    command = ["pbench-list-tools", "--help"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert b"Usage: pbench-list-tools [OPTIONS]" in out
-    assert exitcode == 0
-
-
-def test_list_tool(monkeypatch, agent_config, pbench_run, pbench_cfg):
-    monkeypatch.setenv("_PBENCH_AGENT_CONFIG", str(pbench_cfg))
-    p = pbench_run / "tools-v1-default" / "testhost.example.com"
-    p.mkdir(parents=True)
-    tool = p / "perf"
-    tool.touch()
-
-    command = ["pbench-list-tools"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"default: testhost.example.com ['perf']" in out
-
-    command = ["pbench-list-tools", "--group", "default"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"default: testhost.example.com ['perf']" in out
-
-    command = ["pbench-list-tools", "-n", "perf"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"tool name: perf groups: default" in out
+class Test_list_tool_no_tools_registered:
+    def test_help(self):
+        command = ["pbench-list-tools", "--help"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert b"Usage: pbench-list-tools [OPTIONS]" in out
+        assert exitcode == 0
+
+    def test_no_args(self, monkeypatch, agent_config, pbench_run, pbench_cfg):
```

Comment:
Ugh; these fixtures are very modular and convoluted. I'm surprised there isn't one built on top of `agent_config` that provides the monkeypatched environment variable. There probably should be. And since `agent_config` is already built on top of `pbench_run` and `pbench_cfg`, there's no need to repeat them in the individual tests!

I'd suggest adding the monkeypatch to `agent_config` in the `conftest.py` and then you can simply remove all of the fixtures listed here except `agent_config`, which would simplify all the test cases and make them easier to read.

And, yeah, I'm aware that may be a lot more background rototilling than you'd really intended here, but the fixture dependency structure in these tests is pretty messy.

Reply:
Sure - that should not be a big deal. And since you volunteered, I know where to turn to for help :-)

Reply:
> these fixtures are very modular and convoluted

@dbutenhof, did you omit a "not" from that statement?  (I usually think of modularity, at least when practiced well, as an antidote to convolution....)

Reply:
> > these fixtures are very modular and convoluted
> 
> @dbutenhof, did you omit a "not" from that statement? (I usually think of modularity, at least when practiced well, as an antidote to convolution....)

No, they're convoluted to follow because they're excessively modular; each fixture builds on the previous with a small change, and most if not all of the intermediate fixtures are completely useless except as a building block for the next step. And yet each fixture and each test directly references all of them.

Reply:
Huh.  If each fixture references the previous one, then the test should only need to reference the last...right?

Reply:
@dbutehnof: I left `conftest.py` mostly alone (I just added the `monkeypatch.setenv` to `agent_config`) but eliminated all the extraneous fixtures from the tests.

Reply:
Yes; and each fixture need only reference the one before it in the chain.

Reply:
@webbnh: that's what I did. It seems to work too (miraculously enough).